### PR TITLE
feat(manager/uv): add support for git sources

### PR DESCRIPTION
**Pull request by NextFire**
## Changes

Add support for [uv git sources](https://docs.astral.sh/uv/concepts/dependencies/#git). It's quite the same logic as in #26367.

## Context

Users can define git sources for uv project dependencies in `pyproject.toml`. For instance, in https://github.com/NextFire/renovate-repro-uv-git-rev/blob/main/pyproject.toml, `cowsay` is one tag behind, and `yt-dlp` is a few revisions behind.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Pull request examples: https://github.com/NextFire/renovate-repro-uv-git-rev/pulls

### CODE REVIEW

File: lib/modules/manager/pep621/processors/uv.spec.ts [hunk 1]
```diff
@@ -97,32 +95,100 @@ describe('modules/manager/pep621/processors/uv', () => {
...
           skipReason: 'unsupported-url',
         },
         {
           depName: 'dep5',
+          depType: 'tool.uv.sources',
           skipReason: 'inherited-dependency',
         },
         {
           depName: 'dep6',
-          skipReason: 'invalid-dependency-specification',
```
Comment (secustor) [inline lib/modules/manager/pep621/processors/uv.spec.ts:116]: These `skipReason` shouldn't disappear

File: lib/modules/manager/pep621/processors/uv.ts [hunk 2]
```diff
@@ -175,6 +180,40 @@ export class UvProcessor implements PyProjectProcessor {
   }
 }
 
+function applyGitSource(
+  dep: PackageDependency,
+  depSource: z.infer<typeof UvGitSource>,
```
Comment (secustor) [inline lib/modules/manager/pep621/processors/uv.ts:185]:
```suggestion
  depSource: UvGitSource,
```

Please no inline infers.

File: lib/modules/manager/pep621/schema.ts [hunk 3]
```diff
@@ -35,14 +35,33 @@ const HatchSchema = z.object({
...
-  path: z.string().optional(),
-  url: z.string().optional(),
-  workspace: z.boolean().optional(),
+export const UvGitSource = z.object({
+  git: z.string(),
+  rev: z.string().optional(),
+  tag: z.string().optional(),
+  branch: z.string().optional(),
+});
+
```
Comment (secustor) [inline lib/modules/manager/pep621/schema.ts:44]:
```suggestion
export type UvGitSource = z.infer<typeof UvGitSource>

```

File: lib/modules/manager/pep621/processors/uv.ts [hunk 4]
```diff
@@ -175,6 +180,40 @@ export class UvProcessor implements PyProjectProcessor {
...
+): void {
+  const { git, rev, tag, branch } = depSource;
+  if (tag) {
+    const { source, owner, name } = parseGitUrl(git);
+    const repo = `${owner}/${name}`;
```
Comment (secustor) [inline lib/modules/manager/pep621/processors/uv.ts:190]:
```suggestion
    const { source, full_name: repo } = parseGitUrl(git);
```

File: lib/modules/manager/pep621/utils.ts [hunk 5]
```diff
@@ -18,6 +18,7 @@ export const depTypes = {
   optionalDependencies: 'project.optional-dependencies',
   pdmDevDependencies: 'tool.pdm.dev-dependencies',
   uvDevDependencies: 'tool.uv.dev-dependencies',
+  uvSources: 'tool.uv.sources',
```
Comment (secustor) [inline lib/modules/manager/pep621/utils.ts:21]: This change needs to be reflected in the README

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/pep621/processors/uv.spec.ts (hunk 1 above)
Comment (NextFire) [reply lib/modules/manager/pep621/processors/uv.spec.ts:116]: The [new union](https://github.com/renovatebot/renovate/pull/31928/files#diff-060b5f839d75d04f183bebed7b71c536d693cfb10b610c4f4914a0b6fd7c8e17R59-R64) do not allow the invalid sources specifications [that were previously tested](https://github.com/renovatebot/renovate/pull/31928/files#diff-4e070e322f5eb5bb28ff9cb6c872a7f1c0dff9021ec92f4ae19971d20507f5a9L74-L75) anymore.
(This skip reason also was removed from [the processor](https://github.com/renovatebot/renovate/pull/31928/files#diff-d8da9aafff81cdba1a3495ed391d9536f840192d4d417ee7fea96447ca421c2fL58).)

File: lib/modules/manager/pep621/processors/uv.ts [hunk 6]
```diff
@@ -175,6 +179,36 @@ export class UvProcessor implements PyProjectProcessor {
   }
 }
 
+function applyGitSource(dep: PackageDependency, depSource: UvGitSource): void {
+  const { git, rev, tag, branch } = depSource;
+  if (tag) {
+    const { source, full_name: repo } = parseGitUrl(git);
+    if (source === 'github.com') {
```
Comment (secustor) [inline lib/modules/manager/pep621/processors/uv.ts:186]:
```suggestion
   const platform = detectPlatform(source);
    if (platform === 'github') {
    dep.registryUrls = [`${test.protocol}://${test.source}`];
```
That way self hosted instances will be supported.

Same then for Gitlab.

**Review by secustor — CHANGES_REQUESTED** [review]

File: lib/modules/manager/pep621/processors/uv.ts (hunk 6 above)
Comment (NextFire) [reply lib/modules/manager/pep621/processors/uv.ts:186]: Having a `registryUrls` value will set [`UV_EXTRA_INDEX_URL`](https://github.com/renovatebot/renovate/pull/31928/files/44b41bd4585ea2d809538b935d2a16a99a0ad045#diff-d8da9aafff81cdba1a3495ed391d9536f840192d4d417ee7fea96447ca421c2fR242-R265) and I'm not sure we want that (I'm no more able to resolve some dependencies https://github.com/NextFire/renovate-repro-uv-git-rev/pull/19#issuecomment-2412322530)

File: lib/modules/manager/pep621/processors/uv.spec.ts (hunk 1 above)
Comment (secustor) [reply lib/modules/manager/pep621/processors/uv.spec.ts:116]: So UV basically maps falls back here to PyPi resolution?

File: lib/modules/manager/pep621/processors/uv.ts (hunk 6 above)
Comment (secustor) [reply lib/modules/manager/pep621/processors/uv.ts:186]: UV_EXTRA_INDEX_URL should filter by datasource in that case, based on the 

```typescript
  const pyPiRegistryUrls = deps
    .filter((dep) => dep.datasource === PypiDatasource.id)
    .map((dep) => dep.registryUrls)
    .flat()
  const registryUrls = new Set(pyPiRegistryUrls);
```

File: lib/modules/manager/pep621/processors/uv.spec.ts (hunk 1 above)
Comment (NextFire) [reply lib/modules/manager/pep621/processors/uv.spec.ts:116]: If the source spec is invalid, `uv` will fail to parse it and will not install the project

- `cowsay = { workspace = false }`
```
error: Failed to generate package metadata for `renovate-repro-uv-git-rev==0.1.0 @ virtual+.`
  Caused by: Failed to parse entry for: `cowsay`
  Caused by: `workspace = false` is not yet supported
``` 

- `cowsay = {  }`
```
error: Failed to parse: `pyproject.toml`
  Caused by: TOML parse error at line 10, column 10
   |
10 | cowsay = {  }
   |          ^^^^
expected one of `git`, `url`, `path`, `index`, or `workspace`
```

File: lib/modules/manager/pep621/processors/uv.ts (hunk 6 above)
Comment (NextFire) [reply lib/modules/manager/pep621/processors/uv.ts]: This works, thank you!

**Review by secustor — APPROVED** [review]

File: lib/modules/manager/pep621/processors/uv.spec.ts (hunk 1 above)
Comment (secustor) [reply lib/modules/manager/pep621/processors/uv.spec.ts:116]: I have meant is there isn't any definition in the sources.

Comment (NextFire) [reply lib/modules/manager/pep621/processors/uv.spec.ts:116]: Ah yes, same as `dep1` indeed

**Pull request merged**